### PR TITLE
DBZ-642 Enforce single task per connector

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-642

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java [hunk 1]
```diff
@@ -19,10 +23,20 @@
...
     public static final int DEFAULT_MAX_QUEUE_SIZE = 8192;
     public static final int DEFAULT_MAX_BATCH_SIZE = 2048;
     public static final long DEFAULT_POLL_INTERVAL_MILLIS = 500;
 
+    public static final Field TASKS_MAX = Field.create("tasks.max")
+            .withDisplayName("Maximum tasks per connector")
+            .withType(Type.INT)
+            .withWidth(Width.SHORT)
+            .withImportance(Importance.LOW)
+            .withDescription("Maximum tasks that can be started by connector, only '0' or '1' is allowd")
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java]: s/allowd/allowed/

Does 0 even make sense? Seems fixing it to 1 would be more useful

File: debezium-core/src/main/java/io/debezium/config/Field.java [hunk 2]
```diff
@@ -1123,6 +1123,16 @@ public static int isPositiveInteger(Configuration config, Field field, Validatio
         return 1;
     }
 
+    public static int isZeroOrOne(Configuration config, Field field, ValidationOutput problems) {
+        String value = config.getString(field);
+        if (value == null) return 0;
+        try {
+            if (Integer.parseInt(value) == 0 || Integer.parseInt(value) == 1) return 0;
+        } catch (Throwable e) {}
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/config/Field.java]: Let's use the newly agreed on style for code after closing brackets.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnector.java [hunk 3]
```diff
@@ -100,6 +101,8 @@ public String version() {
     @Override
     public void start(Map<String, String> props) {
         // Validate the configuration ...
+        CommonConnectorConfig.validateConnectConfig(props);
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnector.java]: How about extracting a base class for all the connectors and doing this once in its `start()` method?

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java (hunk 1 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java]: @gunnarmorling You are right, I've checked the Connect code and 1 is the minimum

**Comment by rhauch** [comment]
I'm confused by this issue and PR. It seems strange that you'd limit the number of tasks in the connector configuration, when Connect already allows a user to control the maximum number of tasks. The MySQL and PG connectors can only run a single task, no matter what the maximum number of tasks are. And IIRC MongoDB would compute the number of tasks (for parallel processing) based upon the number of collections and the max number of tasks.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnector.java (hunk 3 above)
Comment (jpechane) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnector.java]: @gunnarmorling done

**Comment by jpechane** [comment]
@rhauch What will happen if a user (and we had such reports) overeagerly configure 4 tasks for MySQL connector? IMHO it will either behave erratically or will fail with a confusin message. As I'd say that we are starting Readers per task not per connector.

**Comment by rhauch** [comment]
@jpechane, the `MySqlConnector.taskConfigs(...)` is responsible for returning the configuration for each task that will be started. It's currently doing nothing, but it should return a list with a single map (task configuration). Same for the PostgreSQL connector. 

The MongoDB connector is [already doing it properly](https://github.com/debezium/debezium/blob/master/debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnector.java#L141-L169) by looking at how much work there is, computing the correct number of tasks, and then defining the configuration for each of those tasks.

**Comment by gunnarmorling** [comment]
Hey @rhauch, the idea is to protect against a configuration of max tasks we actually can't support. I.e. it doesn't make sense to start the MySQL or PG connector with a max.tasks > 1, potentially leaving the user hoping that it'd achieve a better parallelism (this came up recently). So we thought we'd rather fail early on, to make that clear. Would that make sense?

But indeed we missed that the MongoDB connector actually supports more than one task, that had escaped my memory. So indeed this early fail should be done for the two other connectors then only.

**Comment by jpechane** [comment]
@rhauch You are right, thanks for pointing me to the right direction!

**Comment by rhauch** [comment]
@jpechane ah, sorry. comments passed in transit again.

**Comment by jpechane** [comment]
@rhauch @gunnarmorling Guys, how about a little bit more interactive gitter chat? :-)

**Comment by gunnarmorling** [comment]
@rhauch Yes, having re-read the JavaDocs of `taskConfigs()`, I think it's best to leave things as is.

For some reason I was under the wrong assumption that the configured number is authoritative on the number of tasks started, but as you say, the connector is still in control (and we rightly return list with a single element for MySQL and PG, no matter what max task number has been configured). So a higher count is just ignored.

We might log a warning if a higher count than 1 is given, stating that this takes no affect, but it's probably not worth the effort.

Thanks for pointing this out, @rhauch!

**Comment by jpechane** [comment]
Will not be implemented

**Pull request closed without merging**